corehq/ex-submodules/casexml/apps/phone/caselogic.py

```diff
@@ -569,24 +570,53 @@ def get_case_updates(user, last_sync):
     """
     return CaseSyncOperation(user, last_sync)
 
-def filter_cases_modified_elsewhere_since_sync(cases, last_sync):
-    # this function is pretty ugly and is heavily optimized to reduce the number
+
+def filter_cases_modified_elsewhere_since_sync(cases, last_sync_token):
+    """
+    This function takes in a list of unwrapped case dicts and a last_sync token and
+    returns the set of cases that should be applicable to be sent down on top of that
+    sync token.
+
+    This includes:
+
+      1. All cases that were modified since the last sync date by any phone other
+         than the phone that is associated with the sync token.
+      2. All cases that were not on the phone at the time of last sync that are
+         now on the phone.
+    """
+    # todo: this function is pretty ugly and is heavily optimized to reduce the number
     # of queries to couch.
-    if not last_sync:
+    if not last_sync_token:
         return cases
     else:
-        # todo: if this case list is huge i'm guessing this query is pretty expensive
-        case_ids = [case['_id'] for case in cases]
+        # we can start by filtering out our base set of cases to check for only
+        # things that have been modified since we last synced
+        def _is_relevant(case_or_case_state_dict):
+            if case_or_case_state_dict:
+                # only case-like things have this.
+                if 'server_modified_on' in case_or_case_state_dict:
+                    return string_to_utc_datetime(case['server_modified_on']) >= last_sync_token.date
+            # for case states default to always checking for recent updates
+            return True
+
+        recently_modified_case_ids = [case['_id'] for case in cases if _is_relevant(case)]
+        # create a mapping of all cases to sync logs for all cases that were modified
+        # in the appropriate ranges.
+        # todo: this should really have a better way to filter out updates from sync logs
+        # that we already have in a better way.
+        # todo: if this recently modified case list is huge i'm guessing this query is
+        # pretty expensive
         case_log_map = CommCareCase.get_db().view(
             'phone/cases_to_sync_logs',
-            keys=case_ids,
+            keys=recently_modified_case_ids,
```

Comment:
this is the major change, filtering the base list of ids to what has been modified since the last sync if that information is available